### PR TITLE
Wikilink handler interface

### PR DESCRIPTION
**Pull request by kmaasrud**
# Motivation

This PR has a rewrite of how wikilinks are handled, which allows users of the API to add their own handlers to get custom functionality.

# Evidences

None yet

# To do

- [x] Main proof-of-concept interface
- [x] Wikilink preprocessors and renderers
- [ ] New tests for wikilink parsing
- [ ] Revamp `CommonMarkHandler` to be more easily pluggable
- [ ] Write docs

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -2,7 +2,7 @@ extern crate pest;
 #[macro_use]
 extern crate pest_derive;
 
-mod arc;
+mod wikilink;
```
Comment (MatteoJoliveau) [inline src/lib.rs:5]: This needs to be `pub mod wikilink` or `WikilinkHandler` has to be re-exported in order for downstream projects to be able to implement it.

File: src/wikilink/context.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
+use crate::{utils::string::kebab, Note, HashMap};
+use pest::Parser;
+
+pub struct WikilinkContext<'a> {
```
Comment (MatteoJoliveau) [inline src/wikilink/context.rs:4]:
```suggestion
#[derive(Debug)
pub struct WikilinkContext<'a> {
```

This helps with debugging custom handlers.

File: src/lib.rs (hunk 1 above)
Comment (kmaasrud) [reply src/lib.rs:5]: You're right, thanks for pointing that out!

I wrote this up quite quickly today, so there's a lot of stuff like this that needs to be fixed

**Review by MatteoJoliveau — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (MatteoJoliveau) [reply src/lib.rs:5]: No worries, that's what code review is for! 🚀

**Comment by MatteoJoliveau** [comment]
I think there might be something weird happening with aliases, as now the links come out as `<a href="README.html">alias(9, 18)</a>` instead of the proper title. Is there something we must down after wikihandlers have run to get the proper text out?

**Comment by kmaasrud** [comment]
> I think there might be something weird happening with aliases, as now the links come out as `<a href="README.html">alias(9, 18)</a>` instead of the proper title. Is there something we must down after wikihandlers have run to get the proper text out?

Ah fuck... Pest has a weird API to get their rules out as strings 😅 Fixed now

**Comment by MatteoJoliveau** [comment]
It works! 🚀

**Pull request closed without merging**